api/db-cdr/generate-cdr/libproject/genomic_manifests.rb

```diff
@@ -296,41 +465,75 @@ def _process_files_by_manifest(manifest_path, logs_dir, status_verb, num_workers
 # for publishing. For details, see:
 # https://docs.google.com/document/d/1EHw5nisXspJjA9yeZput3W4-vSIcuLBU5dPizTnk1i0/edit
 def stage_files_by_manifest(project, manifest_path, logs_dir, concurrency = GSUTIL_TASK_CONCURRENCY)
+  common = Common.new
   deploy_account = must_get_env_value(project, :publisher_account)
 
+  all_tasks = CSV.read(manifest_path, headers: true, return_headers: false)
+
+  # For now, we support pulling specifically from a ct preprod workspace project as a source.
+  # This scenario is special-cased, since it's where we're doing operational prep of CDR assets.
+  # For publishing in lower environments, i.e. from an arbitrary bucket, just use a normal curationSource.
+  preprod_deploy_account = must_get_env_value("all-of-us-rw-preprod", :publisher_account)
+  preprod_ingest_bucket = must_get_env_value("all-of-us-rw-preprod", :accessTiers)["controlled"][:ingest_cdr_bucket]
+
+  needs_preprod_grant = false
+  if all_tasks.any? { |task| not task["preprod_source_ingest_path"].to_s.empty? }
+    unless ["all-of-us-rw-prod", "all-of-us-rw-preprod"].include? project
+      raise ArgumentError.new("specified a preprod source with project #{project}, this is not allowed")
+    end
+    # preprod -> prod requires a temporary access grant
+    needs_preprod_grant = project == "all-of-us-rw-prod"
+  end
+
+  if needs_preprod_grant
+    common.run_inline(["gsutil", "-i", preprod_deploy_account, "iam", "ch", "serviceAccount:#{deploy_account}:storageAdmin", preprod_ingest_bucket])
+  end
+
   _process_files_by_manifest(
-      manifest_path, File.join([logs_dir, 'stage']), "Staged", concurrency) do |task, wout, werr|
+    all_tasks, File.join([logs_dir, "stage", File.basename(manifest_path, ".csv")]), "Staged", concurrency) do |task, wout, werr|
     ingest_path = task["ingest_path"]
-    [ingest_path, system(
-       "gsutil",
-       "-i",
-       deploy_account,
-       "cp",
-       task["source_path"],
-       ingest_path,
-       :out => wout,
-       :err => werr)]
+
+    unless task["preprod_source_ingest_path"].to_s.empty?
+      # preprod workspace -> (cp) -> preprod ingest -> (mv) -> prod ingest
+      unless system(
+          "gsutil",
+          "-i",
+          preprod_deploy_account,
+          "cp",
+          task["source_path"],
+          task["preprod_source_ingest_path"],
+          :out => wout,
+          :err => werr)
+        next [ingest_path, false]
+      end
+      next [ingest_path, system(
+         "gsutil",
+         "-i",
+         deploy_account,
+         "mv",
+         task["preprod_source_ingest_path"],
+         ingest_path,
+         :out => wout,
+         :err => werr)]
+    end
+    next [ingest_path, system(
```

Comment:
This doesn't need to be a `next` because it's at the end of the loop, right?

Actually are we even using these returned values?

Reply:
the `next` is effectively acting as a `return` from within a block; actually calling return does weird things to the flow control in Ruby.

Reply:
ok thanks